torch/distributed/fsdp/flat_param.py

```diff
@@ -128,8 +128,11 @@ class FlatParameter(nn.Parameter):
         flattened parameter, or the unsharded flattened parameter.
 
     Attributes:
-        _unsharded_size (torch.Size): Unsharded flattened parameter's size
-            (without padding).
+        _unpadded_unsharded_size (torch.Size): Unsharded flattened parameter's
+            size without padding.
+        _padded_unsharded_size (torch.Size): Unsharded flattened parameter's
```

Comment:
Might result in some confusing code in the future if these attributes are being used across FSDP instances where maybe one instance isn't sharded. 

Could we consider setting it for non-sharded strategies and just having it equal to unpadded_unsharded_size since we don't need to pad?

Reply:
I thought about this for a while, and I want to play devil's advocate.

As is, the sharded strategies are handled differently compared to `NO_SHARD` in a few places. It is only in those sharded strategy code paths that `_padded_unsharded_size` is used. I would find it a yellow flag if we are using `_padded_unsharded_size` for both `NO_SHARD` and the sharded strategies; it would be an indication that we may actually want to be using the `_unpadded_unsharded_size` or more broadly the unpadded unsharded flattened parameter.

An example of this would be the old `summon_full_params(offload_to_cpu=True)`. This previously used the *padded* unsharded flattened parameter, but since it was agnostic to whether the sharding strategy was a sharded one or not, we could get away with simply using the *unpadded* unsharded flattened parameter. In other words there is no need to include the padding. (This is what the "generalize prefetching" PR does.)

Given this, it may actually prevent silent errors/mistakes to have `_padded_unsharded_size` only be defined for sharded strategies and hence ensure that we do not inadvertently use the _padded_ unsharded flattened parameter when the _unpadded_ one suffices.